### PR TITLE
load DNN from memory buffer

### LINKED ISSUE #9765: Feature Request : ability to load Caffe/Tensorflow from FileStorage or from similar in-memory
##### System information (version)

- OpenCV =>  3.3
- Operating System / Platform => macOS 10.11
- Compiler => Xcode 8.2.1 / clang-800.0.42.1

##### Detailed description

Like we could load cv::ml::SVM from a cv::FileStorage :

~~~
cv::FileStorage fs;
m_svm = cv::Algorithm::read<cv::ml::SVM>(fs["SVM"]);
~~~

It will look like : `cv::dnn::Net = readNetFromTensorflow(fs["Model"]);`

### PR DESCRIPTION
**Pull request by r2d3**
resolves #9765

### This pullrequest changes

add an overloaded readNetFromTensorflow/readNetFromCaffe functions to read a TF/Caffe Net from data already in memory

### CODE REVIEW

File: modules/dnn/include/opencv2/dnn/dnn.hpp [hunk 1]
```diff
@@ -639,6 +639,11 @@ CV__DNN_EXPERIMENTAL_NS_BEGIN
       */
     CV_EXPORTS_W Net readNetFromTensorflow(const String &model, const String &config = String());
 
+    /** @brief Reads a network model stored in Tensorflow model in memory
+     */
+    CV_EXPORTS_W Net readNetFromTensorflowBuffer(const char *bufferModel, size_t lenModel,
```
Comment (dkurt) [inline modules/dnn/include/opencv2/dnn/dnn.hpp]: Let's keep the same name for it, `readNetFromTensorflow`.

I think we can use brief description like at https://docs.opencv.org/master/d3/d63/classcv_1_1Mat.html#a2ec3402f7d165ca34c7fd6e8498a62ca: `This is an overloaded member function...` where only parameters are described.

Have you tested python / Java bindings for it?

File: modules/dnn/src/tensorflow/tf_io.cpp [hunk 2]
```diff
@@ -52,6 +52,18 @@ bool ReadProtoFromBinaryFileTF(const char* filename, Message* proto) {
     return success;
 }
 
+bool ReadProtoFromBinaryBufferTF(const char* data, size_t len, Message* proto) {
```
Comment (dkurt) [inline modules/dnn/src/tensorflow/tf_io.cpp]: I think it's better to take existing two methods, remove `std::ifstream` creation from them and replace `const char* filename` argument to `std::istream& is` one.
In result, we'll have something like 

```cpp
bool ReadProtoFromTextFileTF(std::istream& s, Message* proto) {
    IstreamInputStream input(&is);
    return google::protobuf::TextFormat::Parse(&input, proto);
}
```
```cpp
bool ReadProtoFromTextFileTF(const char* buf, size_t len, Message* proto) {
    std::istringstream s(std::string(buf, len));
    return ReadProtoFromTextFileTF(s, proto);
}
```
```cpp
bool ReadProtoFromTextFileTF(const char* filename, Message* proto) {
    std::ifstream fs(filename, std::ios::in);
    CHECK(fs.is_open()) << "Can't open \"" << filename << "\"";
    return ReadProtoFromTextFileTF(fs, proto);
}
```

Comment (r2d3) [reply modules/dnn/src/tensorflow/tf_io.cpp]: The problem with std::istringstream s(std::string(buf, len)) is that, in some cases, it creates a copy of the buffer into a std::string.

However ArrayInputStream input(data, len) guarantees that there is no copy
https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.io.zero_copy_stream_impl_lite#ArrayInputStream

Comment (dkurt) [reply modules/dnn/src/tensorflow/tf_io.cpp]: @r2d3, we just vote to reduce a code duplication. There are two methods `ReadProtoFromBinaryFileTF` and `ReadProtoFromBinaryBufferTF` are different only at the beginning for now.

Comment (r2d3) [reply modules/dnn/src/tensorflow/tf_io.cpp]: OK I will reduce code duplication

File: modules/dnn/test/test_tf_importer.cpp [hunk 3]
```diff
@@ -69,20 +69,61 @@ TEST(Test_TensorFlow, inception_accuracy)
     normAssert(ref, out);
 }
 
+static bool readFileInMemory(const string& filename, char** data, size_t* len)
```
Comment (dkurt) [inline modules/dnn/test/test_tf_importer.cpp]: @r2d3, do you mind if we return a content by `std::string`? In example,
```cpp
static bool readFileInMemory(const string& filename, std::string* content) {
  std::ifstream ifs(filename.c_str());
  CV_Assert(ifs.is_open());

  ifs.seekg(0, std::ios::end);
  content->resize(ifs.tellg());
  ifs.seekg(0, std::ios::beg);
  ifs.read(&content->operator[](0), content->size());
}
```

Comment (r2d3) [reply modules/dnn/test/test_tf_importer.cpp]: No, I was thinking about it as it is how I do file loading in personal tests

Comment (r2d3) [reply modules/dnn/test/test_tf_importer.cpp]: Done

File: modules/dnn/src/tensorflow/tf_io.cpp (hunk 2 above)
Comment (r2d3) [reply modules/dnn/src/tensorflow/tf_io.cpp]: I put the common code in `ReadProtoFromBinaryTF`.

I simplified the code in other functions. `raw_input` and `coded_input` are now created on stack to avoid new/delete.

File: modules/dnn/test/test_tf_importer.cpp [hunk 4]
```diff
@@ -69,20 +69,55 @@ TEST(Test_TensorFlow, inception_accuracy)
...
+        string dataModel;
+        ASSERT_TRUE(readFileInMemory(netPath, dataModel));
+
+        string dataConfig;
+        if (hasText)
+            ASSERT_TRUE(readFileInMemory(netConfig, dataConfig));
+
+        net = readNetFromTensorflow(dataModel.c_str(), dataModel.size(),
+                                    dataConfig.c_str(), dataConfig.size());
+        ASSERT_FALSE(net.empty());
```
Comment (dkurt) [inline modules/dnn/test/test_tf_importer.cpp]: I think this check is useful for both ways of importing.

Comment (r2d3) [reply modules/dnn/test/test_tf_importer.cpp]: You are right, test is now done for both ways of importing

**Comment by r2d3** [comment]
Hello @dkurt 

As TensorFlow from memory is now working. I began working on Caffe one. I began moving all the ReadProtoFrom* functions to caffe_io.cpp and using them in tf_io.cpp.

**Comment by r2d3** [comment]
So readNetFromCaffe is now also implemented with a corresponding test (I had to duplicate the `readFileInMemory` function, there is place to put these test utility functions)

I do not know if we could implement the same mechanism for DarkNet and Torch ?!

Regards

File: modules/dnn/include/opencv2/dnn/dnn.hpp [hunk 5]
```diff
@@ -634,11 +634,33 @@ CV__DNN_EXPERIMENTAL_NS_BEGIN
...
 
+    /** @brief Reads a network model stored in Caffe model in memory.
+      * @details This is an overloaded member function, provided for convenience.
+      * It differs from the above function only in what argument(s) it accepts.
+      * @param bufferProto buffer containing the content of the .prototxt file
+      * @param lenProto length of bufferProto
+      * @param bufferModel buffer containing the content of the .caffemodel file
+      * @param lenModel length of bufferModel
+      */
+    CV_EXPORTS_W Net readNetFromCaffe(const char *bufferProto, size_t lenProto,
```
Comment (dkurt) [inline modules/dnn/include/opencv2/dnn/dnn.hpp]: I had a problem with using it in python because of null characters in serialized protocol buffers. I think we should use just `CV_EXPORTS`.

Comment (r2d3) [reply modules/dnn/include/opencv2/dnn/dnn.hpp]: If we use only CV_EXPORTS, the function will not be exposed to Python/Java interface.

I should look at how the binding generator works so the function accept Python str instead of string

Comment (alalek) [reply modules/dnn/include/opencv2/dnn/dnn.hpp]: It can't be exposed, because bindings don't work properly with native C++ pointers (at least they don't support buffers handling).

Comment (r2d3) [reply modules/dnn/include/opencv2/dnn/dnn.hpp]: In `modules/python/src2/hdr_parser.py`, I find a `CV_CARRAY` that seems to be able to wrap arrays. But the corresponding `#define` has disappeared from OpenCV headers.

File: modules/dnn/test/test_caffe_importer.cpp [hunk 6]
```diff
@@ -55,6 +55,43 @@ static std::string _tf(TString filename)
     return (getOpenCVExtraDir() + "/dnn/") + filename;
 }
 
+static bool readFileInMemory(const string& filename, string& content)
```
Comment (dkurt) [inline modules/dnn/test/test_caffe_importer.cpp]: Do you mind if we replace it to `test_common.hpp` to reduce duplicated code? (see `test_tf_importer.cpp`)

Comment (r2d3) [reply modules/dnn/test/test_caffe_importer.cpp]: No problem. I was looking for this kind of common file and did not find it

File: modules/dnn/test/test_caffe_importer.cpp [hunk 7]
```diff
@@ -55,6 +55,43 @@ static std::string _tf(TString filename)
...
+    content.reserve(ifs.tellg());
+    ifs.seekg(0, std::ios::beg);
+
+    content.assign((std::istreambuf_iterator<char>(ifs)),
+                    std::istreambuf_iterator<char>());
+
+    return true;
+}
+
+TEST(Test_Caffe, memory_read)
```
Comment (dkurt) [inline modules/dnn/test/test_caffe_importer.cpp]: I think it'd be better to test an accuracy of one of the models too.

Comment (r2d3) [reply modules/dnn/test/test_caffe_importer.cpp]: in the `memory_read` test case (as I have done for `Test_TensorFlow`) or in a separate test case ?

Comment (dkurt) [reply modules/dnn/test/test_caffe_importer.cpp]: In example, replace `TEST(Reproducibility_AlexNet, Accuracy)` to
```cpp
typedef TestWithParam<tuple<bool> > Reproducibility_AlexNet;
TEST_P(Reproducibility_AlexNet, Accuracy) {
  float readFromMemory = get<0>(GetParam());
  ...
}

INSTANTIATE_TEST_CASE_P(Test_Caffe, Reproducibility_AlexNet, Values(
  Bool()
));
```

Comment (r2d3) [reply modules/dnn/test/test_caffe_importer.cpp]: done

**Comment by dkurt** [comment]
:+1:
@r2d3, thank you! Looks good for me.
